From daf9f5c16d177e0b60818c7211b6eefc0ed5db0c Mon Sep 17 00:00:00 2001
From: Mats Erik Andersson <gnu@gisladisker.se>
Date: Fri, 24 Feb 2017 00:02:14 +0100
Subject: [PATCH 21/64] telnetd: Scrub USER from environment.

Avoid conflicting user identity by forgetting whatever
the serving host considers to be the process owner.
---
 ChangeLog         | 11 +++++++++++
 telnetd/telnetd.c |  8 ++++++++
 2 files changed, 19 insertions(+)

diff --git a/ChangeLog b/ChangeLog
index 7fee820a..7732685a 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,14 @@
+2017-02-23  Mats Erik Andersson  <gnu@gisladisker.se>
+
+	telnetd: Scrub USER from environment.
+	Discard the environment variable USER.  It will later be set
+	properly for autologin, but at least one BSD system passes
+	a preset value when telnetd starts, a value which will cause
+	rejected login when autologin is not in effect.
+
+	* telnetd/telnetd.c (telnetd_setup): Unset environment
+	variable USER before calling getterminaltype().
+
 2017-02-21  Mats Erik Andersson  <gnu@gisladisker.se>
 
 	telnetd: Debugging of mainly line mode options.
diff --git a/telnetd/telnetd.c b/telnetd/telnetd.c
index 3902d77d..69f36a58 100644
--- a/telnetd/telnetd.c
+++ b/telnetd/telnetd.c
@@ -504,6 +504,14 @@ telnetd_setup (int fd)
 
   io_setup ();
 
+  /* Before doing anything related to the identity of the client,
+   * scrub the environment variable USER, since it may be set with
+   * an irrelevant user name at this point.  OpenBSD has been known
+   * to offend at this point with their own inetd.  Any demand for
+   * autologin will get attention in getterminaltype().
+   */
+  unsetenv ("USER");
+
   /* get terminal type. */
   uname[0] = 0;
   level = getterminaltype (uname, sizeof (uname));
-- 
2.28.0.163.g6104cc2f0b6

